src/freenet/node/DarknetPeerNode.java

```diff
@@ -1466,19 +1479,58 @@ public int sendFileOffer(HTTPUploadedFile file, String message) throws IOExcepti
 		return sendFileOffer(fnam, mime, message, data);
 	}
 
+        // handler for sendTextFeed
 	public void handleFproxyN2NTM(SimpleFieldSet fs, int fileNumber) {
 		String text = null;
 		long composedTime = fs.getLong("composedTime", -1);
 		long sentTime = fs.getLong("sentTime", -1);
 		long receivedTime = fs.getLong("receivedTime", -1);
-	  	try {
+		try {
 			text = Base64.decodeUTF8(fs.get("text"));
 		} catch (IllegalBase64Exception e) {
 			Logger.error(this, "Bad Base64 encoding when decoding a N2NTM SimpleFieldSet", e);
 			return;
 		}
-		N2NTMUserAlert userAlert = new N2NTMUserAlert(this, text, fileNumber, composedTime, sentTime, receivedTime);
+		long msgid = fs.getLong("msgid", Long.MAX_VALUE);
+		List<UserAlert> merged = new ArrayList<>();
+		if (msgid != Long.MAX_VALUE) {
+			// merge existing user alerts
+
+                        // TODO: This costs time linear to the number
+			// of known alerts, so receiving a large text
+			// message is quadratic in its size! Worst
+			// case cost for transfer of a single huge
+			// n2ntm (when receiving every second message)
+			// is about 0.5 N*N, with N at most
+			// 128. However even sending a 64kiB file
+			// (DarknetPeerNode.java) does not cause
+			// noticeable load on the receiving system.
+                        synchronized (text) {
```

Comment:
Am I understanding the motivation of this code correctly like that:
Because you didn't write code to intercept split messages to not display a UserAlert until the full message is available the split messages are displayed immediately as partial incomplete UserAlerts and pieced together as more messages arrive?

If yes then this is a big usability issue which I'd consider a blocker:
By exposing the user to randomly cut messages without any apparent reason it forces the user experience to obey the internal technical architecture. It should be precisely the other way around: The internal technical architecture should be forced to comply to produce results which are pleasant to the user = Messages should not be displayed until they are complete.

(Or in any other words: Think about the dozen instant messengers which were popular in the past 20 years - did any of them do this? We'll make it rain bug reports upon ourselves if we ship this :) I'm actually amazed, you invented a new kind of hack here :)

Reply:
This is not "because I did not write" such code, but because this solution is robust against many kinds of failure. Even if we lose a message in between, the user gets two large messages with only the middle part missing.

The synchronization is already fixed (yepp, that was just wrong).